### ISSUE TITLE
Using client behind a proxy?

### ISSUE THREAD

**Issue by rowanwins** [mapbox/mapbox-sdk-js#195]
Hi there,

I'm trying to upload a dataset from behind our corporate proxy by when I `createUploadCredentials()` I appear to get blocked by my corporate proxy, the request eventually times out.

Are you aware of any workarounds?

Plan B - Do it from home :)

**Comment by andrewharvey** [comment]
Hi @rowanwins could you see if this is specific to mapbox-sdk-js or if you have the problem running any node application which uses the network, and also any node application running https://github.com/cujojs/rest which is what this library uses? If it's not specific to this library, I don't think there is much we can do.

**Comment by rowanwins** [comment]
When using [axios](https://github.com/axios/axios) to make requests across the network I can pass in some proxy options which seem to work.

**Comment by xkrys** [comment]
I also can't use the sdk behind a proxy even it https_proxy environment variable is set (linux)

**Comment by fighterii** [comment]
I have the same issue behind a corporate proxy.

Without the proxy everything works as expected.

I stumbled over the following module:
https://github.com/dimebox/rest-node-https-proxy-interceptor

When i edit the client.js on my computer:

```
var restHttpsProxy = require("rest-node-https-proxy-interceptor")
// rest.js client with MIME support
module.exports = function(config) {
  return rest
    .wrap(restHttpsProxy, {
        url: "http://user:password@1.1.1.1:8080/" // you can also use tcp://127.0.0.1:3128
    })
```
The call works as expected.

Maybe you could introduce a new optional option, where the user can specify a proxy and the rest object is wrapped with the proxy-agent.

**Comment by el2ro** [comment]
+1
It would be really nice to have proxy support for mapbox-sdk-js.
I am trying to run mapbox-sdk-js in Node in Docker behind the corporate firewall and it fails.
Even tough my Linux (docker) environment has https_proxy defined, requests will end up timeout due mapbox is not using proxies.

> info: [Mapbox] Reverse geocode error { RequestError: connect ETIMEDOUT xx.xx.xx.xx:443
    at ClientRequest.req.once.err (node_modules/got/index.js:182:22)
    at Object.onceWrapper (events.js:277:13)
    at ClientRequest.emit (events.js:189:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  name: 'RequestError',
  code: 'ETIMEDOUT',
  host: 'api.mapbox.com',
  hostname: 'api.mapbox.com',
  method: 'GET',
  path:
   '/geocoding/v5/mapbox.places/23.68....
  protocol: 'https:',
  url:
   'https://api.mapbox.com/geocoding/v5/mapbox.places/...
 
Using curl to given path in same docker image works as it uses https_proxy env variable.

If I got it right, mapbox-sdk-js node is using "got" package to create http(s) request to Mapbox API. There are some info how proxies could be applied to got library. https://www.npmjs.com/package/got#proxies

It would be really nice if mapbox could look for environment if it has an https_proxy defined and uses that as proxy then.
